lib/sisimai/bite/email/exim.rb

```diff
@@ -0,0 +1,522 @@
+            rcptinhead = mhead['x-failed-recipients'].split(',')
+            rcptinhead.each do |a|
+              # Remove space characters
+              a.lstrip!
+              a.rstrip!
+            end
+            recipients = rcptinhead.size
+
+            rcptinhead.each do |e|
+              # Insert each recipient address into @$dscontents
+              dscontents[-1]['recipient'] = e
+              next if dscontents.size == recipients
+              dscontents << Sisimai::Bite.DELIVERYSTATUS
+            end
+          end
+        end
+        return nil if recipients.zero?
+
+        if mhead['received'].size > 0
+          # Get the name of local MTA
+          # Received: from marutamachi.example.org (c192128.example.net [192.0.2.128])
+          if cv = mhead['received'][-1].match(/from[ \t]([^ ]+)/)
+            localhost0 = cv[1]
+          end
+        end
+
+        require 'sisimai/string'
+        require 'sisimai/smtp/reply'
+        require 'sisimai/smtp/status'
+
+        dscontents.map do |e|
+          # Set default values if each value is empty.
+          e['agent']   = self.smtpagent
+          e['lhost'] ||= localhost0
+
+          unless e['diagnosis']
+            # Empty Diagnostic-Code: or error message
+            if boundary00.size > 0
+              # --NNNNNNNNNN-eximdsn-MMMMMMMMMM
+              # Content-type: message/delivery-status
+              #
+              # Reporting-MTA: dns; the.local.host.name
+              #
+              # Action: failed
+              # Final-Recipient: rfc822;/a/b/c
+              # Status: 5.0.0
+              #
+              # Action: failed
+              # Final-Recipient: rfc822;|/p/q/r
+              # Status: 5.0.0
+              e['diagnosis'] = dscontents[0]['diagnosis'] || ''
+              e['spec']    ||= dscontents[0]['spec']
+
+              if dscontents[0]['alterrors'] && dscontents[0]['alterrors'].size > 0
+                # The value of "alterrors" is also copied
+                e['alterrors'] = dscontents[0]['alterrors']
+              end
+            end
+          end
+
+          if e['alterrors'] && e['alterrors'].size > 0
+            # Copy alternative error message
+            if e['diagnosis'].nil? || e['diagnosis'].empty?
+              e['diagnosis'] ||= e['alterrors']
+            end
+
+            if e['diagnosis'] =~ /\A[-]+/ || e['diagnosis'] =~ /__\z/
+              # Override the value of diagnostic code message
+              e['diagnosis'] = e['alterrors'] if e['alterrors'].size > 0
+
+            else
+              # Check the both value and try to match
+              if e['diagnosis'].size < e['alterrors'].size
+                # Check the value of alterrors
+                rxdiagnosis = %r/#{e['diagnosis']}/i
+                if e['alterrors'] =~ rxdiagnosis
+                  # Override the value of diagnostic code message because
+                  # the value of alterrors includes the value of diagnosis.
```

Comment:
__[rubocop]__
* Performance/EndWith: Use `String#end_with?` instead of a regex match anchored to the end of the string. [:ref](https://github.com/JuanitoFatas/fast-ruby#stringmatch-vs-stringstart_withstringend_with-code-start-code-end)

<!-- Comment by [SideCI](https://sideci.com) -->